src/Microsoft.AspNet.OData.Shared/Formatter/Serialization/ODataResourceSerializer.cs

```diff
@@ -634,6 +638,105 @@ public virtual ODataResource CreateResource(SelectExpandNode selectExpandNode, R
+                {
+                    if (resource.InstanceAnnotations == null)
+                    {
+                        resource.InstanceAnnotations = new List<ODataInstanceAnnotation>();
+                    }
+
+                    foreach (var annote in instanceAnnotationsDictionary[string.Empty])
+                    {
+                        IEdmTypeReference edmTypeReference = resourceContext.SerializerContext.GetEdmType(annote.Value,
+                           annote.Value.GetType());
+
+                        ODataEdmTypeSerializer edmTypeSerializer;
+
+                        if (edmTypeReference.IsComplex() || edmTypeReference.IsEntity() || edmTypeReference.IsCollection())
+                        {
+                            if (edmTypeReference.IsCollection())
+                            {
+                                edmTypeSerializer = new ODataCollectionSerializer(SerializerProvider,true);
+                            }
+                            else
+                            {
+                                edmTypeSerializer = new ODataComplexSerializer(SerializerProvider);
+                            }                            
+                        }
+                        else
+                        {
+                            edmTypeSerializer = SerializerProvider.GetEdmTypeSerializer(edmTypeReference);
+                        }
+                                                
+                        if (edmTypeSerializer != null)
+                        {
+                           ODataValue annoteVal = edmTypeSerializer.CreateODataValue(annote.Value, edmTypeReference, resourceContext.SerializerContext);
+
+                            if (annoteVal != null)
+                            {
+                                resource.InstanceAnnotations.Add(new ODataInstanceAnnotation(annote.Key, annoteVal));
+                            }
+                        }                       
+                    }
+                }
+                
+                foreach(var property in resource.Properties)
+                {
+                    string propertyName = property.Name;
+
+                    if (instanceAnnotationsDictionary.ContainsKey(propertyName))
+                    {
+                        if (property.InstanceAnnotations == null)
+                        {
+                            property.InstanceAnnotations = new List<ODataInstanceAnnotation>();
+                        }
+
+                        foreach (var annote in instanceAnnotationsDictionary[propertyName])
+                        {
+                            IEdmTypeReference edmTypeReference = resourceContext.SerializerContext.GetEdmType(annote.Value,
+                            annote.Value.GetType());
+                       
+                            ODataEdmTypeSerializer edmTypeSerializer = SerializerProvider.GetEdmTypeSerializer(edmTypeReference);
```

Comment:
> ODataEdmTypeSerializer edmTypeSerializer = SerializerProvider.GetEdmTypeSerializer(edmTypeReference); [](start = 27, length = 102)

Why is the logic for getting a serializer different for an annotation on a property (here) than it is for an annotation on the type (lines 679-693)?  I would expect that the logic would be the same, and you would have a common private method called for the type as well as for each property to serialize any annotations.

Reply:
#Resolved